Add Flask Instrumentation fixes

ocelotl opened the pull request:
Fixes #599 

This is just a separate PR to keep the Flask changes from blocking the instrumentation enhancements.

lzchen commented:
Reopening up the discussion.
The changes here are good. It address @mauriciovasquezbernal [comments](https://github.com/open-telemetry/opentelemetry-python/pull/579#discussion_r410284499) on your previous PR. I'm still wondering if patching the class instead of the app itself is the way to go.

mauriciovasquezbernal commented:
I think it is actually not addressing my comments yet. 
I have some concerns about they way `instrument` is implemented here and how it should be used https://github.com/open-telemetry/opentelemetry-python/pull/579#discussion_r411471658.

I exposed an alternative to this here https://github.com/open-telemetry/opentelemetry-python/pull/579#pullrequestreview-395464048.

ocelotl commented:
> I think it is actually not addressing my comments yet.
> I have some concerns about they way `instrument` is implemented here and how it should be used [#579 (comment)](https://github.com/open-telemetry/opentelemetry-python/pull/579#discussion_r411471658).
> 
> I exposed an alternative to this here [#579 (review)](https://github.com/open-telemetry/opentelemetry-python/pull/579#pullrequestreview-395464048).

Adding @mauriciovasquezbernal comments from past reviews:

> Even if I don't completely agree with having an ABC and so on, I think the changes to `BaseInstrumentor` in the last iteration are good. Making it a singleton and allowing arbitrary arguments helps.
> 
> However I think the changes to the Flask are not that good, the proposed way for users to enable instrumentation is rather complicated. I still think we should have `instrument` and `uninstrument` methods that are simple to use (just a single call, few parameters and not return value) that are used by both, the automatic instrumentation command and the users. I'm aware that in the users case this function could not work in all the cases, for instance if you call it too late in the code after importing some modules, but I think that with correct documentation and warnings they would be very useful.
> 
> For the programmatic case, we could use a much simpler approach, just expose and let the user interact with `InstrumentedFlask`.
> 
> ```
> from opentelemetry.ext.flask import InstrumentedFlask as Flask
> app = Flask(...) # app will be instrumented
> ```
> 
> or
> 
> ```
> from flask import Flask
> from opentelemetry.ext.flask import InstrumentedFlask
> app1 = Flask(...) # this won't be instrument
> app2 = InstrumentedFlask(...) # this will be 
> ```
> 
> The `instrument` method should be used for automatic instrumentation and when the user want's to instrument everything in that framework, the later with some restrictions, like enforcing a proper order in the imports.
> 
> ```
> from opentelemetry.ext.flask import FlaskInstrumentor
> FlaskInstrumentor().instrument()
> 
> from flask import Flask # needed to do after
> 
> app = Flask(...) # instrumented
> ```
> 
> The following case won't work, but we could print a warning to the user that the flask module is already imported and it might not work.
> 
> ```
> from flask import Flask
> from opentelemetry.ext.flask import FlaskInstrumentor
> FlaskInstrumentor().instrument()
> 
> app = Flask(...) # uninstrumented, reference to Flask was taken before instrumenting
> ```

mauriciovasquezbernal commented on docs/ext/django/django.rst:
```diff
@@ -0,0 +1,7 @@
+OpenTelemetry Django Instrumentation
```
Probably this file should not be in this PR.

ocelotl replied:
Removed

mauriciovasquezbernal commented on ext/opentelemetry-ext-flask/src/opentelemetry/ext/flask/__init__.py:
```diff
@@ -29,12 +29,13 @@
 
     app = Flask(__name__)
 
+    FlaskInstrumentor().instrument(app=app)
```
```suggestion
    FlaskInstrumentor().instrument_app(app)
```

?

ocelotl replied:
Weird, I am sure I changed this... fixing.

mauriciovasquezbernal commented on ext/opentelemetry-ext-flask/src/opentelemetry/ext/flask/__init__.py:
```diff
@@ -56,113 +57,145 @@ def hello():
+    def instrument_app(self, app):
+        # FIXME remove this protection once some mechanism in the
+        # BaseInstrumentor class is available for these methods.
+        if not self._is_instrumented:
```
What about making this decision considering only information present in `app`?, for instance checking if `app.wsgi_app` is `_rewrapped_app` or something similar?

I think using `_is_instrumented`  is not correct because it refers to the whole framework and not to the single app instance.

ocelotl replied:
Fixed

mauriciovasquezbernal commented on ext/opentelemetry-ext-flask/src/opentelemetry/ext/flask/__init__.py:
```diff
@@ -56,113 +57,145 @@ def hello():
+
+            app.before_request(_before_request)
+            app.teardown_request(_teardown_request)
+            self._is_instrumented = True
```
Does it mean that I cannot instrument two different apps?


```
instrument_app(app1)
instrument_app(app2) <- will fail because is_instrumented was set by the previous call.
```

ocelotl replied:
Fixed, moved the check to the app itself.

mauriciovasquezbernal commented on ext/opentelemetry-ext-flask/tests/test_automatic.py:
```diff
@@ -0,0 +1,98 @@
+        self.assertEqual(span_list[0].attributes, expected_attrs)
+
+
+if __name__ == "__main__":
```
Is it required on a test file?

ocelotl replied:
Removed

mauriciovasquezbernal commented on ext/opentelemetry-ext-flask/tests/test_programmatic.py:
```diff
@@ -0,0 +1,83 @@
+class TestProgrammatic(WsgiTestBase, InstrumentationTest):
+    def setUp(self):
+        # No instrumentation code is here because it is present in the
+        # conftest.py file next to this file.
```
Is this comment still valid?

ocelotl replied:
Fixed

mauriciovasquezbernal commented on ext/opentelemetry-ext-flask/tests/test_programmatic.py:
```diff
@@ -0,0 +1,83 @@
+        FlaskInstrumentor().uninstrument_app(self.app)
+        disable(NOTSET)
+
+    def test_unonstrument(self):
```
```suggestion
    def test_uninstrument(self):
```

mauriciovasquezbernal commented on ext/opentelemetry-ext-flask/tests/test_programmatic.py:
```diff
@@ -0,0 +1,83 @@
+        self.assertEqual(span_list[0].attributes, expected_attrs)
+
+
+if __name__ == "__main__":
```
Same here.

ocelotl replied:
It is an usual testing convention but not necessary since we use pytest, removing.

mauriciovasquezbernal commented on opentelemetry-auto-instrumentation/src/opentelemetry/auto_instrumentation/instrumentor.py:
```diff
@@ -52,7 +52,9 @@ def instrument(self, **kwargs):
             return result
 
-        _LOG.warning("Attempting to instrument while already instrumented")
+        _LOG.warning(
```
Is this change intentional?

ocelotl replied:
Must have slipped from other PR, reverting.

mauriciovasquezbernal commented on ext/opentelemetry-ext-flask/tests/test_automatic.py:
```diff
@@ -0,0 +1,98 @@
+# the instrumentor in a way that mimics how it would be called with the
+# opentelemetry-auto-instrumentation command. This does not mean that the
+# instrumentor should be used in this way in end user applications. For those
+# cases, FlaskInstrumentor.instrument(app=app) should be used.
```
```suggestion
# cases, FlaskInstrumentor.instrument_app(app) should be used.
```

?

ocelotl replied:
That comment does not apply anymore, removed.

ocelotl commented:
Addressed all your comments, @mauriciovasquezbernal, please review again when you have a moment. @lzchen, these changes address your previous concern about doing instrumentation by replacing the `Flask` class, please take a look.

Thanks!

lzchen commented:
This looks good. I'm curious as to whether we can merge the two methods of auto-instrumentation and programmatic instrumentation but still have it patch an `Flask` app instance instead of the class.
Similar to what we are doing in [SQLAlchemy](https://github.com/open-telemetry/opentelemetry-python/blob/master/ext/opentelemetry-ext-sqlalchemy/src/opentelemetry/ext/sqlalchemy/__init__.py#L72).

ocelotl commented:
> This looks good. I'm curious as to whether we can merge the two methods of auto-instrumentation and programmatic instrumentation but still have it patch an `Flask` app instance instead of the class.
> Similar to what we are doing in [SQLAlchemy](https://github.com/open-telemetry/opentelemetry-python/blob/master/ext/opentelemetry-ext-sqlalchemy/src/opentelemetry/ext/sqlalchemy/__init__.py#L72).

It is possible but we preferred not to do so. We prefer not to have a method whose actions are different depending on the optional attributes it gets, we rather use `**kwargs` in `_instrument` only to pass configuration attributes that will override the configuration options read from the `Configuration` object.

mauriciovasquezbernal commented on ext/opentelemetry-ext-flask/tests/base_test.py:
```diff
@@ -91,7 +59,7 @@ def assert_environ():
         self.assertEqual(nonstring_keys, set())
 
-    def test_simple(self):
+    def test_simple_uninstrument(self):
```
I don't understand the name of this test. `simple_uninstrument` but there are not any calls to uninstrument in the test.

ocelotl replied:
Fixed

mauriciovasquezbernal commented on ext/opentelemetry-ext-flask/tests/test_automatic.py:
```diff
@@ -0,0 +1,98 @@
+
+        self.app = flask.Flask(__name__)
+
+        def hello_endpoint(helloid):
+            if helloid == 500:
+                raise ValueError(":-(")
+            return "Hello: " + str(helloid)
+
+        def excluded_endpoint():
+            return "excluded"
+
+        def excluded2_endpoint():
+            return "excluded2"
+
+        self.app.route("/hello/<int:helloid>")(hello_endpoint)
+        self.app.route("/excluded/<int:helloid>")(hello_endpoint)
+        self.app.route("/excluded")(excluded_endpoint)
+        self.app.route("/excluded2")(excluded2_endpoint)
+
+        self.client = Client(self.app, BaseResponse)
```
All this code is duplicated in test_programmatic. What about reducing this by creating a method in `InstrumentationTest` that performs this initialization? or a similar approach?

ocelotl replied:
Fixed

mauriciovasquezbernal commented on ext/opentelemetry-ext-flask/tests/test_automatic.py:
```diff
@@ -0,0 +1,98 @@
+        self.client = Client(self.app, BaseResponse)
+
+    def tearDown(self):
+        disable(WARNING)
```
`super().tearDown()`?

ocelotl replied:
Fixed

mauriciovasquezbernal commented on ext/opentelemetry-ext-flask/tests/test_automatic.py:
```diff
@@ -0,0 +1,98 @@
+        FlaskInstrumentor().uninstrument()
+        self.app = flask.Flask(__name__)
+
+        def hello_endpoint(helloid):
+            if helloid == 500:
+                raise ValueError(":-(")
+            return "Hello: " + str(helloid)
```
Would it be possible to define `hello_endpoint` as a static method on the class to reuse it?

ocelotl replied:
Fixed

mauriciovasquezbernal commented on ext/opentelemetry-ext-flask/tests/test_automatic.py:
```diff
@@ -0,0 +1,98 @@
+        self.assertEqual([b"Hello: 123"], list(resp.response))
+        span_list = self.memory_exporter.get_finished_spans()
+        self.assertEqual(len(span_list), 1)
+        self.assertEqual(span_list[0].name, "hello_endpoint")
+        self.assertEqual(span_list[0].kind, trace.SpanKind.SERVER)
+        self.assertEqual(span_list[0].attributes, expected_attrs)
```
These values are generated before the call to `uninstrument`, why to check them again?
Just checking that no new spans were generated should be enough.

ocelotl replied:
Fixed

mauriciovasquezbernal commented on ext/opentelemetry-ext-flask/tests/test_programmatic.py:
```diff
@@ -0,0 +1,87 @@
+        disable(NOTSET)
+
+    def test_uninstrument(self):
+        expected_attrs = expected_attributes(
+            {"http.target": "/hello/123", "http.route": "/hello/<int:helloid>"}
+        )
+        resp = self.client.get("/hello/123")
+        self.assertEqual(200, resp.status_code)
+        self.assertEqual([b"Hello: 123"], list(resp.response))
+        span_list = self.memory_exporter.get_finished_spans()
+        self.assertEqual(len(span_list), 1)
+        self.assertEqual(span_list[0].name, "hello_endpoint")
+        self.assertEqual(span_list[0].kind, trace.SpanKind.SERVER)
+        self.assertEqual(span_list[0].attributes, expected_attrs)
```
I think there are other tests already checking the attributes and so on. What do you think about just checking that a span was generated?

ocelotl replied:
Fixed

mauriciovasquezbernal commented:
btw, would you mind moving the entry point definitions to setup.cfg file to keep things uniform with other instrumentations?

ocelotl commented on docs/getting-started.rst:
```diff
@@ -104,6 +104,7 @@ Configure exporters to emit spans elsewhere
 
 The example above does emit information about all spans, but the output is a bit hard to read.
 In common cases, you would instead *export* this data to an application performance monitoring backend, to be visualized and queried.
+oHEA
```
Fixed

ocelotl commented on ext/opentelemetry-ext-flask/tests/test_automatic.py:
```diff
@@ -0,0 +1,98 @@
+from opentelemetry.test.wsgitestutil import WsgiTestBase
+
+# pylint: disable=import-error
+from .base_test import InstrumentationTest, expected_attributes
```
Tried that already, `opentelemetry.ext.flask.tests` is not an importable path.

ocelotl commented on ext/opentelemetry-ext-flask/tests/test_programmatic.py:
```diff
@@ -0,0 +1,76 @@
+        self.client = Client(self.app, BaseResponse)
+
+    def tearDown(self):
+        disable(WARNING)
```
Nice, using it now.

ocelotl commented:
All your comments have been addressed, @mauriciovasquezbernal, thanks for the review!

mauriciovasquezbernal commented on ext/opentelemetry-ext-flask/tests/test_programmatic.py:
```diff
@@ -0,0 +1,56 @@
+        super().tearDown()
+        with self.disable_logging():
+            FlaskInstrumentor().uninstrument_app(self.app)
+
```
Is it worth to check that calling (un)-instrument twice doesn't cause any issue?

mauriciovasquezbernal commented on docs/examples/opentelemetry-example-app/src/opentelemetry_example_app/flask_example.py:
```diff
@@ -34,15 +34,15 @@
 
-
 app = flask.Flask(__name__)
 
```
Not related to this PR, but is this example missing `app.run()`, how am I supposed to run it?

toumorokoshi commented on ext/opentelemetry-ext-flask/src/opentelemetry/ext/flask/__init__.py:
```diff
@@ -176,18 +179,50 @@ def _disable_trace(url):
+            app.wsgi_app = app._original_wsgi_app
+
+            # FIXME add support for other Flask blueprints that are not None
+            app.before_request_funcs[None].remove(_before_request)
```
have you considered using the instrumented flag to turn off and on behavior, like the pymongo instrumentation? 

Or you could remove by value by comparing the id of the function:

```
for k, v in before_request_funcs.items():
    if v is _before_request:
        del before_request_funcs[k]
```

Although the latter feels more error prone as you are relying on the data structure remaining the same across different versions of flask. The checking of the _is_instrumented value feels more robust here.
